scipy/linalg/_matfuncs_sqrtm.h

```diff
@@ -0,0 +1,186 @@
+    if ((PyArray_TYPE(ap_Am) != NPY_FLOAT64) &&
+         (PyArray_TYPE(ap_Am) != NPY_FLOAT32) &&
+         (PyArray_TYPE(ap_Am) != NPY_COMPLEX64) &&
+         (PyArray_TYPE(ap_Am) != NPY_COMPLEX128)
+       )
+    {
+        PYERR(sqrtm_error, "Input must be nD array of type "
+              "float32, float64, complex64, or complex128.");
+    }
+
+    int ndim = PyArray_NDIM(ap_Am);              // Number of dimensions
+    npy_intp* shape = PyArray_SHAPE(ap_Am);      // Array shape
+    npy_intp n = shape[ndim - 1];                // Slice size
+    int input_type = PyArray_TYPE(ap_Am);        // Data type enum value
+    // Compare last two dimensions for squareness
+    if (n != shape[ndim - 2])
+    {
+        PYERR(sqrtm_error, "Last two dimensions of the input must be the same.");
+    }
+
+    // Create the output array with the same shape as the input with twice the
+    // number of entries to accomodate for potential complex-valued data from
+    // real data which will be cast to complex e.g., "ret.asview(complex128)"
+    // Example, (3, 3) -> (18), (4, 5, 5) -> (4, 50)
+    npy_intp ret_dims = 1;
+    for (int i = 0; i < ndim; i++) { ret_dims *= shape[i]; }
+    if ((input_type == NPY_FLOAT32) || (input_type == NPY_FLOAT64))
+    {
+        ret_dims *= 2;
+    }
+
+    if (PyArray_TYPE(ap_Am) == NPY_FLOAT32)
+    {
+        mem_ret = malloc(ret_dims*sizeof(float));
+        matrix_squareroot_s(ap_Am, (float*)mem_ret, &isIllconditioned, &isSingular, &info, &isComplex);
+
+    } else if (PyArray_TYPE(ap_Am) == NPY_FLOAT64) {
+
+        mem_ret = malloc(ret_dims*sizeof(double));
+        matrix_squareroot_d(ap_Am, (double*)mem_ret, &isIllconditioned, &isSingular, &info, &isComplex);
+
+    } else if (PyArray_TYPE(ap_Am) == NPY_COMPLEX64) {
+
+        mem_ret = malloc(ret_dims*sizeof(SCIPY_C));
+        matrix_squareroot_c(ap_Am, (SCIPY_C*)mem_ret, &isIllconditioned, &isSingular, &info, &isComplex);
+        isComplex = 1;
+
+    } else if (PyArray_TYPE(ap_Am) == NPY_COMPLEX128) {
+
+        mem_ret = malloc(ret_dims*sizeof(SCIPY_Z));
+        matrix_squareroot_z(ap_Am, (SCIPY_Z*)mem_ret, &isIllconditioned, &isSingular, &info, &isComplex);
+        isComplex = 1;
+
+    } else {
+        PYERR(sqrtm_error, "Unsupported input data type to sqrtm C function.");
+    }
+
+    if (info < 0)
+    {
+        // Internal failure memory or LAPACK error, fail and return the error code in info
+        free(mem_ret);
+        Py_INCREF(Py_None);
+        return Py_BuildValue("Niii", Py_None, isIllconditioned, isSingular, info);
+    }
+
+    if (!isComplex)
+    {
+        // Input and output is real, truncate the extra space at the end
+        npy_intp new_size = (ret_dims/2)*(input_type == NPY_FLOAT32 ? sizeof(float) : sizeof(double));
+        void* mem_ret_half = realloc(mem_ret, new_size);
+        // Quite unlikely but still allowed to fail
+        if (!mem_ret_half) { PYERR(sqrtm_error, "Memory reallocation failed."); }
```

Comment:
need to `free` the original `mem_ret` here?

Reply:
I am not sure. I was under the impression that we are quitting the program so it will return `mem_ret` but let me free it explicitly anyways. Good catch nevertheless.